lib/redis/cluster.rb

```diff
@@ -0,0 +1,272 @@
+      when 'client'   then send_client_command(command, &block)
+      when 'cluster'  then send_cluster_command(command, &block)
+      when 'readonly', 'readwrite', 'shutdown'
+        raise OrchestrationCommandNotSupported, cmd
+      when 'memory'   then send_memory_command(command, &block)
+      when 'script'   then send_script_command(command, &block)
+      when 'pubsub'   then send_pubsub_command(command, &block)
+      when 'exec'     then @node.call_master(command, &block).reject(&:empty?)
+      when 'discard', 'multi', 'unwatch'
+        @node.call_master(command, &block).first
+      else
+        node = assign_node(command)
+        try_send(node, :call, command, &block)
+      end
+    end
+
+    def send_config_command(command, &block)
+      case command[1].to_s.downcase
+      when 'resetstat', 'rewrite', 'set'
+        @node.call_all(command, &block).first
+      else assign_node(command).call(command, &block)
+      end
+    end
+
+    def send_memory_command(command, &block)
+      case command[1].to_s.downcase
+      when 'stats' then @node.call_all(command, &block)
+      when 'purge' then @node.call_all(command, &block).first
+      else assign_node(command).call(command, &block)
+      end
+    end
+
+    def send_client_command(command, &block)
+      case command[1].to_s.downcase
+      when 'list' then @node.call_all(command, &block).flatten
+      when 'pause', 'reply', 'setname'
+        @node.call_all(command, &block).first
+      else assign_node(command).call(command, &block)
+      end
+    end
+
+    def send_cluster_command(command, &block)
+      subcommand = command[1].to_s.downcase
+      case subcommand
+      when 'addslots', 'delslots', 'failover', 'forget', 'meet', 'replicate',
+           'reset', 'set-config-epoch', 'setslot'
+        raise OrchestrationCommandNotSupported, 'cluster', subcommand
+      when 'saveconfig' then @node.call_all(command, &block).first
+      else assign_node(command).call(command, &block)
+      end
+    end
+
+    def send_script_command(command, &block)
+      case command[1].to_s.downcase
+      when 'debug', 'kill'
+        @node.call_all(command, &block).first
+      when 'flush', 'load'
+        @node.call_master(command, &block).first
+      else assign_node(command).call(command, &block)
+      end
+    end
+
+    def send_pubsub_command(command, &block)
+      case command[1].to_s.downcase
+      when 'channels' then @node.call_all(command, &block).flatten.uniq.sort
+      when 'numsub'
+        @node.call_all(command, &block).reject(&:empty?).map { |e| Hash[*e] }
+             .reduce({}) { |a, e| a.merge(e) { |_, v1, v2| v1 + v2 } }
+      when 'numpat' then @node.call_all(command, &block).reduce(:+)
+      else assign_node(command).call(command, &block)
+      end
+    end
+
+    # @see https://redis.io/topics/cluster-spec#redirection-and-resharding
+    #   Redirection and resharding
+    def try_send(node, method_name, *args, ttl: RETRY_COUNT, &block)
```

Comment:
Maybe it's just me but TTL makes me thing of a duration, not a number of retries. It's true that it's used in this sense for TCP packets (TTL as number of hops), but in Redis that term is used for a number of seconds before expiration.

So it's a tiny bit confusing IMO.